File: taxitips-api/worker/src/taxiRelevance.js
/**
 * Rankar störningar efter nytta för taxiförare.
 *
 * Princip: visa bara när folk faktiskt kan behöva skjuts —
 * inställda tåg, stora störningar, ersättningsbuss, olycka, m.m.
 * Ignorera stationsservice (hiss, toalett, cykel, …).
 */

const { findHubsInText } = require("./hubs");
const { alertInSkane, placeLooksSkane } = require("./skane");

/** Station/info utan taxinytta — även om rubriken nämner en station. */
const NOISE_RE =
  /(hiss(en|ar|arna)?|rulltrapp|rullstol|toalett|cykel|cyklar|wifi|wi-?fi|biljettautomat|biljettmaskin|entr[eé]|assistans|assistent|ledsag|plattformsavvisare|platsbrist|kort tåg|nya spår)/i;

/** Allvarliga kollektivstörningar som brukar ge taxibehov. */
const SERIOUS_RE =
  /\b(inställd|inställda|ställs in|inga avgångar|ingen trafik|trafikstopp|stopp i (trafiken|tågtrafiken|busstrafiken)|totalt stopp|stora störningar|stora förseningar|ersättningsbuss|ersättningstrafik|strejk|nedrivning|strömavbrott|växelfel|signalproblem|tågtrafik (står|stoppad|inställd)|alla (pågatåg|öresundståg|tåg) (är )?inställda|avgång(ar)? inställd|inställd avgång|tåg(en)? (går|kör) inte|banan (är )?avstängd)\b/i;

/** Medel — kan ge efterfrågan men inte alltid “kör hit nu”. */
const MEDIUM_RE =
  /\b(försening|förseningar|minskad (service|trafik)|tågbyte|enkelspårsdrift|hastighetsnedsättning|banarbete som påverkar|förväntas bli (sen|försenad))\b/i;

const CITY_RE =
  /\b(Malmö|Lund|Helsingborg|Kristianstad|Landskrona|Trelleborg|Ystad|Eslöv|Höör|Hässleholm|Ängelholm|Simrishamn|Staffanstorp|Kävlinge|Hyllie|Triangeln|Lomma|Vellinge|Höganäs|Osby|Sjöbo|Svedala|Burlöv|Bromölla|Perstorp|Örkelljunga|Bjuv|Åstorp|Klippan)\b/gi;

const SKANE_ROAD_HINT =
  /\b(malmö|lund|helsingborg|kristianstad|landskrona|trelleborg|ystad|eslöv|hässleholm|ängelholm|simrishamn|kävlinge|hyllie|vellinge|höganäs|lomma|staffanstorp|bromölla|skåne|e22|e6|e65|väg 11|väg 108)\b/i;

function textOf(alert) {
  return `${alert.header || ""}\n${alert.description || ""}\n${alert.cause || ""}\n${alert.effect || ""}`;
}

function placesFrom(alert) {
  const text = `${alert.header || ""} ${alert.description || ""}`;
  const places = [];
  const hubs = findHubsInText(text);
  for (const hub of hubs) {
    if (!places.some((x) => x.toLowerCase() === hub.name.toLowerCase())) {
      places.push(hub.name);
    }
  }
  for (const m of text.matchAll(CITY_RE)) {
    const p = m[0];
    const coveredByHub = hubs.some(
      (h) => h.city.toLowerCase() === p.toLowerCase() || h.name.toLowerCase() === p.toLowerCase()
    );
    if (coveredByHub && hubs.length) continue;
    if (!places.some((x) => x.toLowerCase() === p.toLowerCase())) places.push(p);
  }
  return places.filter((p) => placeLooksSkane(p) || /köpenhamn|cph|kastrup/i.test(String(p)));
}

function ignoreResult(why, alert) {
  return {
    score: 0,
    level: "ignore",
    why,
    places: placesFrom(alert),
    driverHint: null,
    hubs: [],
  };
}

function isNoiseOnly(text) {
  if (!NOISE_RE.test(text)) return false;
  // Hiss + inställda tåg i samma text → behåll allvarliga delen
  return !SERIOUS_RE.test(text);
}

function hasSeriousEffect(alert) {
  const effect = String(alert.effect || "").toLowerCase();
  return (
    effect.includes("ingen service") ||
    effect.includes("inställd") ||
    effect.includes("stoppad") ||
    effect.includes("stora försening") ||
    effect.includes("minskad service")
  );
}

function scoreAlert(alert) {
  const header = (alert.header || "").trim();
  const text = textOf(alert);
  const textLower = text.toLowerCase();

  if (isNoiseOnly(text)) {
    return ignoreResult("Stationsservice (hiss/toalett/cykel m.m.) — ingen taxinytta", alert);
  }

  if (/^stängd hållplats$/i.test(header) || /^hållplats /i.test(header)) {
    if (!SERIOUS_RE.test(text)) {
      return ignoreResult("Enstaka hållplatsinfo", alert);
    }
  }

  if (/^trafikinformation$/i.test(header) && !SERIOUS_RE.test(text) && !MEDIUM_RE.test(text)) {
    return ignoreResult("Allmän trafikinfo", alert);
  }

  // Tillfällig körväg / hållplatsflytt utan inställda tåg → inte “kör hit”
  if (
    /tillfällig (körväg|hållplats)|hänvisas till/i.test(text) &&
    !SERIOUS_RE.test(text) &&
    !/inställd/i.test(text)
  ) {
    return ignoreResult("Omledning/hållplatsflytt — låg taxinytta", alert);
  }

  const hubs = findHubsInText(text);
  const places = placesFrom(alert);
  const placeStr = places.length ? places.join(", ") : "Skåne";
  const hubName = hubs[0]?.name;

  const serious =
    SERIOUS_RE.test(text) ||
    (hasSeriousEffect(alert) &&
      (textLower.includes("tåg") ||
        textLower.includes("påga") ||
        textLower.includes("öresund") ||
        textLower.includes("buss") ||
        hubs.length > 0));

  const mediumish = MEDIUM_RE.test(text) || String(alert.effect || "").toLowerCase().includes("försening");

  // "Tekniskt fel" ensamt (t.ex. hiss) ska aldrig bli high
  const technicalOnly =
    /tekniskt fel/i.test(String(alert.cause || "")) && !serious && !mediumish;

  if (technicalOnly || (!serious && !mediumish && !hasSeriousEffect(alert))) {
    // Orsak/effekt utan allvarliga nyckelord → oftast brus
    if (!serious) {
      return ignoreResult("Ingen allvarlig trafikstörning", alert);
    }
  }

  let score = 0;
  const reasons = [];

  if (serious) {
    score = 70;
    reasons.push("allvarlig störning");
    if (/inställd|ställs in|inga avgångar|ingen trafik/i.test(text)) {
      score = 85;
      reasons.push("inställd/stopp");
    }
    if (/ersättningsbuss|ersättningstrafik/i.test(text)) {
      score += 5;
      reasons.push("ersättningsbuss");
    }
  } else if (mediumish) {
    score = 35;
    reasons.push("försening/påverkan");
  } else {
    return ignoreResult("Otillräcklig taxirelevans", alert);
  }

  if (hubs.length) {
    score += Math.min(12, Math.round((hubs[0].weight || 0) * 0.25));
    reasons.push(`station: ${hubs[0].name}`);
  }

  let level = "ignore";
  if (score >= 60 && serious) level = "high";
  else if (score >= 30) level = "medium";
  else if (score > 0) level = "low";

  // Hub-boost får aldrig ensamt lyfta till high utan allvarlig störning
  if (!serious && level === "high") level = "medium";

  let driverHint = null;
  if (level === "high") {
    driverHint = hubName
      ? `Station ${hubName}: resenärer behöver skjuts — kör hit. ${header}`
      : `Ökad taxieftefrågan trolig i ${placeStr}. Kollektivtrafik störd: ${header}`;
  } else if (level === "medium") {
    driverHint = hubName
      ? `Kolla ${hubName}: ${header}`
      : `Möjlig ökad efterfrågan i ${placeStr}: ${header}`;
  }

  return {
    score,
    level,
    why: reasons.slice(0, 4).join(", ") || "låg relevans",
    places,
    hubs: hubs.map((h) => h.name),
    driverHint,
    // Exposed so mode-aware severity tiering (scoring.js) can reuse the same
    // serious/mediumish signals instead of re-deriving them with a second regex pass.
    serious,
    mediumish,
  };
}

function scoreRoadAlert(alert) {
  const type = String(alert.cause || alert.header || "").toLowerCase();
  const text = `${alert.header || ""} ${alert.description || ""} ${alert.effect || ""}`.toLowerCase();
  let places = Array.isArray(alert.areas)
    ? alert.areas.filter((p) => p && !String(p).startsWith("Linjer") && !String(p).startsWith("Hållplatser"))
    : [];
  const hubs = findHubsInText(text);
  for (const hub of hubs) {
    if (!places.some((x) => x.toLowerCase() === hub.name.toLowerCase())) {
      places = [hub.name, ...places];
    }
  }

  const inSkaneSphere = SKANE_ROAD_HINT.test(text) || places.some((p) => SKANE_ROAD_HINT.test(String(p)));
  const isAccident = type.includes("olycka") || text.includes("olycka");
  const isFullClosure =
    text.includes("vägen avstäng") ||
    text.includes("helt avstäng") ||
    type.includes("avstängning") ||
    type.includes("avstangning");
  const isRoadwork = type.includes("vägarbete") || text.includes("beläggningsarbete") || text.includes("vägarbete");
  const isQueue = type.includes("kö") || text.includes("kövarning") || text.includes("köbildning");

  // Planerat nattarbete / beläggning långt bort → ingen taxisignal
  if (isRoadwork && !isAccident && !inSkaneSphere) {
    return {
      score: 0,
      level: "ignore",
      why: "Vägarbete utanför relevant område",
      places,
      hubs: hubs.map((h) => h.name),
      driverHint: null,
    };
  }

  if (isRoadwork && !isAccident && text.includes("kl 19") && text.includes("05:00")) {
    return {
      score: 0,
      level: "ignore",
      why: "Nattligt vägarbete — låg taxinytta",
      places,
      hubs: hubs.map((h) => h.name),
      driverHint: null,
    };
  }

  let level = "ignore";
  let score = 0;

  if (isAccident && inSkaneSphere) {
    level = "high";
    score = 75;
  } else if (isAccident) {
    level = "medium";
    score = 40;
  } else if (isFullClosure && inSkaneSphere && !isRoadwork) {
    level = "high";
    score = 65;
  } else if (isQueue && inSkaneSphere) {
    level = "medium";
    score = 40;
  } else if (isFullClosure && inSkaneSphere && isRoadwork) {
    // Avstängd väg i Skåne p.g.a. arbete — bevaka, inte “kör hit”
    level = "medium";
    score = 32;
  } else if (isRoadwork && inSkaneSphere) {
    level = "low";
    score = 15;
  } else {
    return {
      score: 0,
      level: "ignore",
      why: "Väginfo utan taxirelevans",
      places,
      hubs: hubs.map((h) => h.name),
      driverHint: null,
    };
  }

  // Vägarbete får aldrig bli high
  if (isRoadwork && !isAccident && level === "high") {
    level = "medium";
    score = Math.min(score, 40);
  }

  const place = places[0] || hubs[0]?.name || "vägen";
  return {
    score,
    level,
    why: isAccident ? "olycka" : isRoadwork ? "vägarbete" : "väg",
    places,
    hubs: hubs.map((h) => h.name),
    driverHint:
      level === "high"
        ? `${alert.cause || "Väg"} nära ${place}: omväg + mer efterfrågan.`
        : level === "medium"
          ? `${alert.cause || "Väginfo"} i ${place} — bra att känna till.`
          : null,
  };
}

function isRoadAlert(alert) {
  return (
    alert?.sourceKind === "road" ||
    alert?.source_kind === "road" ||
    String(alert?.id || "").startsWith("tv:")
  );
}

function enrichAlert(alert) {
  if (isRoadAlert(alert)) {
    const scored = { ...alert, sourceKind: "road", taxi: scoreRoadAlert(alert) };
    if (!alertInSkane(scored) && scored.taxi.level !== "ignore") {
      return {
        ...scored,
        taxi: {
          ...scored.taxi,
          score: 0,
          level: "ignore",
          why: "Utanför Skåne",
          driverHint: null,
        },
      };
    }
    return scored;
  }
  const taxi = scoreAlert(alert);
  const enriched = { ...alert, taxi };
  if (!alertInSkane(enriched) && taxi.level !== "ignore") {
    return {
      ...enriched,
      taxi: {
        ...taxi,
        score: 0,
        level: "ignore",
        why: "Utanför Skåne",
        driverHint: null,
      },
    };
  }
  return enriched;
}

function filterForTaxi(alerts, { minLevel = "medium" } = {}) {
  const order = { high: 3, medium: 2, low: 1, ignore: 0 };
  const min = order[minLevel] ?? 2;
  return alerts
    .map(enrichAlert)
    .filter((a) => (order[a.taxi.level] || 0) >= min)
    .sort((a, b) => b.taxi.score - a.taxi.score || b.lastSeenAt - a.lastSeenAt);
}

function isTaxiNotifyWorthy(alert) {
  const taxi = alert.taxi || (isRoadAlert(alert) ? scoreRoadAlert(alert) : scoreAlert(alert));
  return taxi.level === "high";
}

module.exports = {
  scoreAlert,
  scoreRoadAlert,
  enrichAlert,
  filterForTaxi,
  isTaxiNotifyWorthy,
  isRoadAlert,
};

const h3 = require("h3-js");

function calculateDemandSignal(alert) {
  let h3_index = null;
  // Get lat/lon from alert (from raw payload or parsed)
  const lat = alert.lat || alert.payload?.lat;
  const lon = alert.lon || alert.payload?.lon;
  
  if (lat && lon) {
    h3_index = h3.latLngToCell(lat, lon, 8);
  }

  let demand_score = alert.taxi?.score || 0;
  let reasons = alert.taxi?.why ? alert.taxi.why.split(',').map(s => s.trim()) : [];
  
  const text = (alert.header + " " + alert.description).toLowerCase();
  const isTrain = text.includes("tåg") || text.includes("påga") || text.includes("öresund");
  
  // Last train risk: late night check
  // NOTE: raw source alerts carry active_to (ms epoch), not endsAt -- this never matched
  // before, so the +30 "Last Train Risk" boost has never actually applied.
  const end = alert.active_to ? new Date(alert.active_to) : null;
  if (isTrain && end) {
    const hours = end.getHours();
    // Late night / early morning (e.g. 22:00 - 04:00)
    if (hours >= 22 || hours <= 4) {
      demand_score = Math.min(100, demand_score + 30);
      reasons.push("Last Train Risk");
    }
  }

  return { h3_index, demand_score, reasons };
}

module.exports.calculateDemandSignal = calculateDemandSignal;

File: taxitips-backend/core/taxi_relevance.py
"""
Rankar störningar efter nytta för taxiförare.

Port av worker/src/taxiRelevance.js -- bara den logik som faktiskt matas av
en källa som finns i Django-portens Fas 1 (Trafiklab, textbaserade
transitlarm). Medvetet UTELÄMNAT mot originalet:

- scoreRoadAlert: ingen vägkälla är portad än, inget matar den.
- calculateDemandSignal (h3/"Last Train Risk"): dött kod i originalet --
  bifogad EFTER modulens egna module.exports, läser ett fält
  ("Last Train Risk") som filens egen kommentar säger aldrig triggade.
  poll_rail.py satte redan precedenset att inte porta detta (h3_index="").
- filterForTaxi/isTaxiNotifyWorthy: konsumeras av ett push/notis-steg som
  inte finns i Django än.

Princip: visa bara när folk faktiskt kan behöva skjuts -- inställda tåg,
stora störningar, ersättningsbuss, olycka, m.m. Ignorera stationsservice
(hiss, toalett, cykel, ...).
"""

from __future__ import annotations

import re

from core.geo import find_hubs_in_text, resolve_place_coords
from core.market import (
    alert_in_market,
    is_national_scope,
    place_looks_skane,
    road_text_looks_skane,
)

# Station/info utan taxinytta -- även om rubriken nämner en station.
NOISE_RE = re.compile(
    r"(hiss(en|ar|arna)?|rulltrapp|rullstol|toalett|cykel|cyklar|wifi|wi-?fi|"
    r"biljettautomat|biljettmaskin|entr[eé]|assistans|assistent|ledsag|"
    r"plattformsavvisare|platsbrist|kort tåg|nya spår)",
    re.IGNORECASE,
)

_SV = "a-zà-öø-ÿ0-9"

# Allvarliga kollektivstörningar som brukar ge taxibehov.
SERIOUS_RE = re.compile(
    rf"(?<![{_SV}])(inställd|inställt|inställda|ställs in|inga avgångar|"
    r"ingen trafik|trafikstopp|stopp i (trafiken|tågtrafiken|busstrafiken)|"
    r"totalt stopp|stora störningar|stora förseningar|ersättningsbuss|"
    r"ersättningstrafik|strejk|nedrivning|strömavbrott|växelfel|"
    r"signalproblem|tågtrafik (står|stoppad|inställd)|"
    r"alla (pågatåg|öresundståg|tåg) (är )?inställda|avgång(ar)? inställd|"
    r"inställd avgång|tåg(en)? (går|kör) inte|banan (är )?avstängd)"
    rf"(?![{_SV}])",
    re.IGNORECASE,
)

# Medel -- kan ge efterfrågan men inte alltid "kör hit nu".
MEDIUM_RE = re.compile(
    r"\b(försening|förseningar|minskad (service|trafik)|tågbyte|"
    r"enkelspårsdrift|hastighetsnedsättning|banarbete som påverkar|"
    r"förväntas bli (sen|försenad))\b",
    re.IGNORECASE,
)

CITY_RE = re.compile(
    rf"(?<![{_SV}])(Malmö|Lund|Helsingborg|Kristianstad|Landskrona|Trelleborg|"
    r"Ystad|Eslöv|Höör|Hässleholm|Ängelholm|Simrishamn|Staffanstorp|"
    r"Kävlinge|Hyllie|Triangeln|Lomma|Vellinge|Höganäs|Osby|Sjöbo|Svedala|"
    r"Burlöv|Bromölla|Perstorp|Örkelljunga|Bjuv|Åstorp|Klippan|Stockholm|"
    r"Solna|Södertälje|Nacka|Sundbyberg|Täby|Norrtälje|Uppsala|Enköping|"
    r"Göteborg|Mölndal|Kungsbacka|Borås|Trollhättan|Uddevalla|Skövde|"
    r"Linköping|Norrköping|Motala|Jönköping|Nässjö|Värnamo|Kalmar|"
    r"Oskarshamn|Västervik|Nybro|Karlstad|Kristinehamn|Arvika|Örebro|"
    r"Karlskoga|Västerås|Köping|Eskilstuna|Nyköping|Falun|Borlänge|Mora|"
    r"Gävle|Sandviken|Hudiksvall|Sundsvall|Härnösand|Örnsköldsvik|"
    r"Östersund|Umeå|Skellefteå|Luleå|Piteå|Kiruna|Visby|Karlskrona|"
    r"Karlshamn|Varberg|Halmstad|Växjö|Älmhult)"
    rf"(?![{_SV}])",
    re.IGNORECASE,
)

_COPENHAGEN_RE = re.compile(r"köpenhamn|cph|kastrup", re.IGNORECASE)
_TILLFALLIG_RE = re.compile(r"tillfällig (körväg|hållplats)|hänvisas till", re.IGNORECASE)
# inställ[dt]: "inställd" (en-ord: buss, avgång) och "inställt" (ett-ord:
# tåg -- "tåget är inställt"). Bara "inställd" missade alla ett-ords-former,
# upptäckt via compensation.py:s tester ("Tåg ... inställt").
_INSTALLD_STOPP_RE = re.compile(r"inställ[dt]|ställs in|inga avgångar|ingen trafik", re.IGNORECASE)
_ERSATTNING_RE = re.compile(r"ersättningsbuss|ersättningstrafik", re.IGNORECASE)
_TEKNISKT_FEL_RE = re.compile(r"tekniskt fel", re.IGNORECASE)
_STANGD_HALLPLATS_RE = re.compile(r"^stängd hållplats$", re.IGNORECASE)
_HALLPLATS_RE = re.compile(r"^hållplats ", re.IGNORECASE)
_TRAFIKINFO_RE = re.compile(r"^trafikinformation$", re.IGNORECASE)
_INSTALLD_RE = re.compile(r"inställ[dt]", re.IGNORECASE)


def _text_of(alert: dict) -> str:
    return "\n".join([
        alert.get("header") or "", alert.get("description") or "",
        alert.get("cause") or "", alert.get("effect") or "",
    ])


def places_from(alert: dict) -> list[str]:
    text = f"{alert.get('header') or ''} {alert.get('description') or ''}"
    places: list[str] = []
    hubs = find_hubs_in_text(text)
    for hub in hubs:
        if not any(x.lower() == hub["name"].lower() for x in places):
            places.append(hub["name"])

    for m in CITY_RE.finditer(text):
        p = m.group(0)
        covered_by_hub = any(
            h["city"].lower() == p.lower() or h["name"].lower() == p.lower() for h in hubs
        )
        if covered_by_hub and hubs:
            continue
        if not any(x.lower() == p.lower() for x in places):
            places.append(p)

    return [
        p for p in places
        if resolve_place_coords(p) is not None or place_looks_skane(p) or _COPENHAGEN_RE.search(str(p))
    ]


def _ignore_result(why: str, alert: dict) -> dict:
    return {
        "score": 0, "level": "ignore", "why": why,
        "places": places_from(alert), "driver_hint": None, "hubs": [],
    }


def is_noise_only(text: str) -> bool:
    if not NOISE_RE.search(text):
        return False
    return not SERIOUS_RE.search(text)


def has_serious_effect(alert: dict) -> bool:
    effect = str(alert.get("effect") or "").lower()
    return any(
        s in effect
        for s in ("ingen service", "inställd", "stoppad", "stora försening", "minskad service")
    )


def score_alert(alert: dict) -> dict:
    header = (alert.get("header") or "").strip()
    text = _text_of(alert)
    text_lower = text.lower()

    if is_noise_only(text):
        return _ignore_result("Stationsservice (hiss/toalett/cykel m.m.) — ingen taxinytta", alert)

    if _STANGD_HALLPLATS_RE.match(header) or _HALLPLATS_RE.match(header):
        if not SERIOUS_RE.search(text):
            return _ignore_result("Enstaka hållplatsinfo", alert)

    if _TRAFIKINFO_RE.match(header) and not SERIOUS_RE.search(text) and not MEDIUM_RE.search(text):
        return _ignore_result("Allmän trafikinfo", alert)

    if _TILLFALLIG_RE.search(text) and not SERIOUS_RE.search(text) and not _INSTALLD_RE.search(text):
        return _ignore_result("Omledning/hållplatsflytt — låg taxinytta", alert)

    hubs = find_hubs_in_text(text)
    places = places_from(alert)
    place_str = ", ".join(places) if places else ("Okänd plats" if is_national_scope() else "Skåne")
    hub_name = hubs[0]["name"] if hubs else None

    serious = bool(SERIOUS_RE.search(text)) or (
        has_serious_effect(alert)
        and (
            "tåg" in text_lower or "påga" in text_lower or "öresund" in text_lower
            or "buss" in text_lower or bool(hubs)
        )
    )

    mediumish = bool(MEDIUM_RE.search(text)) or "försening" in str(alert.get("effect") or "").lower()

    # "Tekniskt fel" ensamt (t.ex. hiss) ska aldrig bli high.
    technical_only = bool(_TEKNISKT_FEL_RE.search(str(alert.get("cause") or ""))) and not serious and not mediumish

    if technical_only or (not serious and not mediumish and not has_serious_effect(alert)):
        if not serious:
            return _ignore_result("Ingen allvarlig trafikstörning", alert)

    score = 0
    reasons: list[str] = []

    if serious:
        score = 70
        reasons.append("allvarlig störning")
        if _INSTALLD_STOPP_RE.search(text):
            score = 85
            reasons.append("inställd/stopp")
        if _ERSATTNING_RE.search(text):
            score += 5
            reasons.append("ersättningsbuss")
    elif mediumish:
        score = 35
        reasons.append("försening/påverkan")
    else:
        return _ignore_result("Otillräcklig taxirelevans", alert)

    if hubs:
        score += min(12, round((hubs[0].get("weight") or 0) * 0.25))
        reasons.append(f"station: {hubs[0]['name']}")

    level = "ignore"
    if score >= 60 and serious:
        level = "high"
    elif score >= 30:
        level = "medium"
    elif score > 0:
        level = "low"

    # Hub-boost får aldrig ensamt lyfta till high utan allvarlig störning.
    if not serious and level == "high":
        level = "medium"

    driver_hint = None
    if level == "high":
        driver_hint = (
            f"Station {hub_name}: resenärer behöver skjuts — kör hit. {header}"
            if hub_name else
            f"Ökad taxieftefrågan trolig i {place_str}. Kollektivtrafik störd: {header}"
        )
    elif level == "medium":
        driver_hint = f"Kolla {hub_name}: {header}" if hub_name else f"Möjlig ökad efterfrågan i {place_str}: {header}"

    return {
        "score": score,
        "level": level,
        "why": ", ".join(reasons[:4]) or "låg relevans",
        "places": places,
        "hubs": [h["name"] for h in hubs],
        "driver_hint": driver_hint,
        # Exponerat så mode-medveten allvarlighetsgradering (text_scoring.py)
        # kan återanvända samma serious/mediumish-signaler i stället för att
        # räkna om dem med ett andra regexpass.
        "serious": serious,
        "mediumish": mediumish,
    }


def is_road_alert(alert: dict | None) -> bool:
    if not alert:
        return False
    return (
        alert.get("source_kind") == "road"
        or str(alert.get("id") or "").startswith("tv:")
    )



def score_road_alert(alert: dict) -> dict:
    """
    Port av taxiRelevance.js:s scoreRoadAlert.

    Taket är 15 poäng, med avsikt. En olycka, avstängning eller kö försenar
    dem som REDAN sitter i bil -- den strandar ingen fotgängare så som ett
    inställt tåg gör, och ingen lämnar sin bil mitt i en kö för att ta taxi.
    Det enda verkliga taxifallet är en avstängning så allvarlig att en
    busslinje ställs in, och den störningen kommer som ett eget
    kollektivtrafiklarm -- alltså skulle en högre vägpoäng dubbelräkna den.

    Väginformation är därför sammanhang för en förare som redan är på väg
    ("räkna med omväg"), aldrig ett skäl att köra någonstans.
    """
    type_text = str(alert.get("cause") or alert.get("header") or "").lower()
    text = " ".join(
        str(alert.get(k) or "") for k in ("header", "description", "effect")
    ).lower()

    places = [
        p
        for p in (alert.get("areas") or [])
        if p and not str(p).startswith("Linjer") and not str(p).startswith("Hållplatser")
    ]
    hubs = find_hubs_in_text(text)
    for hub in hubs:
        if not any(str(x).lower() == hub["name"].lower() for x in places):
            places = [hub["name"], *places]

    in_market = road_text_looks_skane(text, places)
    is_accident = "olycka" in type_text or "olycka" in text
    is_full_closure = (
        "vägen avstäng" in text
        or "helt avstäng" in text
        or "avstängning" in type_text
        or "avstangning" in type_text
    )
    is_roadwork = (
        "vägarbete" in type_text or "beläggningsarbete" in text or "vägarbete" in text
    )
    is_queue = "kö" in type_text or "kövarning" in text or "köbildning" in text

    def ignored(why: str) -> dict:
        return {
            "score": 0, "level": "ignore", "why": why, "places": places,
            "hubs": [h["name"] for h in hubs], "driver_hint": None,
            "serious": False, "mediumish": False,
        }

    # Planerat nattarbete eller beläggning långt bort ger ingen taxisignal.
    if is_roadwork and not is_accident and not in_market:
        return ignored("Vägarbete utanför relevant område")
    if is_roadwork and not is_accident and "kl 19" in text and "05:00" in text:
        return ignored("Nattligt vägarbete — låg taxinytta")

    if is_accident and in_market:
        score, level = 15, "low"
    elif is_accident:
        score, level = 8, "low"
    elif is_full_closure and in_market and not is_roadwork:
        score, level = 15, "low"
    elif is_queue and in_market:
        score, level = 10, "low"
    elif is_full_closure and in_market and is_roadwork:
        score, level = 8, "low"
    elif is_roadwork and in_market:
        score, level = 5, "low"
    else:
        return ignored("Väginfo utan taxirelevans")

    place = places[0] if places else (hubs[0]["name"] if hubs else "vägen")
    return {
        "score": score,
        "level": level,
        "why": "olycka" if is_accident else "vägarbete" if is_roadwork else "väg",
        "places": places,
        "hubs": [h["name"] for h in hubs],
        "driver_hint": (
            f"{alert.get('cause') or 'Väginfo'} i {place} — bra att känna till, "
            "men skapar inte taxibehov i sig."
        ),
        "serious": False,
        "mediumish": False,
    }

def enrich_alert(alert: dict) -> dict:
    """
    Port of taxiRelevance.js's enrichAlert: scores the alert, then re-checks
    alert_in_market() and force-zeroes to "ignore" if the alert names
    another region -- the out-of-market suppression must run AFTER
    scoring (it needs taxi["places"]), not folded into score_alert itself.

    Returns the (possibly force-ignored) taxi dict -- not merged with
    `alert`, unlike the JS original's object-spread style; callers already
    have `alert` and pass both to classify_transit_alert().
    """
    if is_road_alert(alert):
        return score_road_alert(alert)

    taxi = score_alert(alert)
    if not alert_in_market(alert, taxi) and taxi["level"] != "ignore":
        return {**taxi, "score": 0, "level": "ignore", "why": "Utanför marknaden", "driver_hint": None}
    return taxi
